wrapper/src/main/java/software/amazon/jdbc/DataSourceConnectionProvider.java

```diff
@@ -157,6 +163,73 @@ public Connection connect(
     return conn;
   }
 
+  protected Connection openConnection(
+      final @NonNull DataSource ds,
+      final @NonNull String protocol,
+      final @NonNull TargetDriverDialect targetDriverDialect,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props)
+      throws SQLException {
+
+    try {
+      return ds.getConnection();
+
+    } catch (Throwable throwable) {
+
+      if (!PropertyDefinition.ENABLE_GREEN_NODE_REPLACEMENT.getBoolean(props)) {
+        throw throwable;
+      }
+
+      UnknownHostException unknownHostException = null;
+      int maxDepth = 100;
+      Throwable loopThrowable = throwable;
+      while (--maxDepth > 0 && loopThrowable != null) {
+        if (loopThrowable instanceof UnknownHostException) {
+          unknownHostException = (UnknownHostException) loopThrowable;
+          break;
+        }
+        loopThrowable = loopThrowable.getCause();
+      }
+
+      if (unknownHostException == null) {
+        throw throwable;
+      }
+
+      if (!this.rdsUtils.isRdsDns(hostSpec.getHost()) || !this.rdsUtils.isGreenInstance(hostSpec.getHost())) {
+        throw throwable;
+      }
```

Comment:
Does it make sense to check the dns first and fail early before going through potentially 100 exceptions?

Reply:
I'm not sure. Checking DNS upfront takes time. Checking when exception thrown is handling error. Maybe we will make such change later as part of optimization.